### ISSUE TITLE
False positive invalid-method-override on overload when self is parent with reduced constrained type

### ISSUE THREAD

**Issue by MeGaGiGaGon** [astral-sh/ty#3921]
### Summary

I think this should be a fine override. Since the overloads differ on the constrained generic, only the `int` overload applies in `AlwaysInt`, so I shouldn't need to write code for the impossible `None` case in `AlwaysInt`.
https://play.ty.dev/431af941-283f-4a90-9d75-701f9e3d8c4c
```py
from abc import ABC, abstractmethod
from typing import overload, override


class Abstract[T, P: (int, None)](ABC):
    @overload
    def foo(self: Abstract[T, int]) -> int: ...
    @overload
    def foo(self: Abstract[T, None]) -> None: ...
    @abstractmethod
    def foo(self) -> int | None: ...

# Expected: no error
# Mypy, pyrefly: no error
# ty, pyright: Error invalid-method-override
class AlwaysInt[T](Abstract[T, int]):
    @override
    def foo(self) -> int:
        return 1
```

### Version

ty 0.0.55 (42869400d 2026-06-26) playground ce7a42e18

**Comment by carljm** [comment]
Thanks for the report! This occurs in your example because `P` is unused in the body of `Abstract`, making it inferred as bivariant, which means `Abstract[T, int]` and `Abstract[T, None]` are the same type, which means we can't eliminate either overload. Adding `marker: P` to the body of `Abstract` (the simplest way to force `P` to invariance instead of bivariance) also eliminates the `invalid-method-override` error (as does a method `def get_p(self) -> P: ...`).

We have a PR to eliminate bivariance and fall back to covariance instead, which would fix this.
